libs/serialization/src/exception_ptr.cpp

```diff
@@ -0,0 +1,311 @@
+                throw_line_ = *line;
+        }
+
+        // figure out concrete underlying exception type
+        try
+        {
+            std::rethrow_exception(ep);
+        }
+        catch (hpx::thread_interrupted const&)
+        {
+            type = hpx::util::hpx_thread_interrupted_exception;
+            what = "hpx::thread_interrupted";
+            err_value = hpx::thread_cancelled;
+        }
+        catch (hpx::exception const& e)
+        {
+            type = hpx::util::hpx_exception;
+            what = e.what();
+            err_value = e.get_error();
+        }
+        catch (boost::system::system_error const& e)
+        {
+            type = hpx::util::boost_system_error;
+            what = e.what();
+            err_value = e.code().value();
+            err_message = e.code().message();
+        }
+        catch (std::runtime_error const& e)
+        {
+            type = hpx::util::std_runtime_error;
+            what = e.what();
+        }
+        catch (std::invalid_argument const& e)
+        {
+            type = hpx::util::std_invalid_argument;
+            what = e.what();
+        }
+        catch (std::out_of_range const& e)
+        {
+            type = hpx::util::std_out_of_range;
+            what = e.what();
+        }
+        catch (std::logic_error const& e)
+        {
+            type = hpx::util::std_logic_error;
+            what = e.what();
+        }
+        catch (std::bad_alloc const& e)
+        {
+            type = hpx::util::std_bad_alloc;
+            what = e.what();
+        }
+        catch (std::bad_cast const& e)
+        {
+            type = hpx::util::std_bad_cast;
+            what = e.what();
+        }
+        catch (std::bad_typeid const& e)
+        {
+            type = hpx::util::std_bad_typeid;
+            what = e.what();
+        }
+        catch (std::bad_exception const& e)
+        {
+            type = hpx::util::std_bad_exception;
+            what = e.what();
+        }
+        catch (std::exception const& e)
+        {
+            type = hpx::util::std_exception;
+            what = e.what();
+        }
+#if BOOST_ASIO_HAS_BOOST_THROW_EXCEPTION != 0
+        catch (boost::exception const& e)
+        {
+            type = hpx::util::boost_exception;
+            what = boost::diagnostic_information(e);
+        }
+#endif
+        catch (...)
+        {
+            type = hpx::util::unknown_exception;
+            what = "unknown exception";
+        }
+
+        // clang-format off
+        ar & type & what & throw_function_ & throw_file_ & throw_line_;
+        // clang-format on
+
+        if (hpx::util::hpx_exception == type)
+        {
+            // clang-format off
+            ar & err_value;
+            // clang-format on
+        }
+        else if (hpx::util::boost_system_error == type)
+        {
+            // clang-format off
+            ar & err_value & err_message;
+            // clang-format on
+        }
+    }
```

Comment:
I think the body of this function should be encapsulated and installed as the default `custom_exception_handler`. If no custom handler is installed we should call `std::terminate` (as this would be a bug).

Reply:
Also, may I suggest to return the previously installed custom handler from the `set_...` functions. This would allow for the user to chain handlers if necessary.

Reply:
I can certainly add that, but I'd like to emphasize that I have not intended for this to be user-facing. This is mainly to resolve the internal dependencies that we have.

Reply:
Ok, fair point.